src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1819 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file zero_elimination.cc
+ * \brief Transform tensors in such a way as to eliminate summation over zeros.
+ */
+#include "zero_elimination.h"
+
+#include <tvm/api_registry.h>
+#include <tvm/ir_functor_ext.h>
+#include <tvm/ir.h>
+#include <tvm/ir_mutator.h>
+#include <tvm/ir_pass.h>
+#include <tvm/ir_visitor.h>
+#include <tvm/operation.h>
+
+#include <algorithm>
+#include <map>
+#include <set>
+#include <utility>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "arithmetic/ModulusRemainder.h"
+#include "../op/op_util.h"
+
+namespace tvm {
+namespace ir {
+
+using HalideIR::Internal::gcd;
+using HalideIR::Internal::lcm;
+
+struct ExprLess {
+    bool operator()(const Expr& l, const Expr& r) const {
+      return Compare(l, r) < 0;
+    }
+};
+
+struct ExprEq {
+    bool operator()(const Expr& l, const Expr& r) const {
+      return Compare(l, r) == 0;
+    }
+};
+
+// Merge two maps, prefer the right one on conflict
+template <class K, class V>
+Map<K, V> Merge(Map<K, V> original, const Map<K, V>& update) {
+  for (const auto& p : update) {
+    original.Set(p.first, p.second);
+  }
+  return std::move(original);
+}
+
+// Concatenate two arrays
+template <class T>
+Array<T> Concat(Array<T> a, const Array<T>& b) {
+  for (const auto& x : b) {
+    a.push_back(x);
+  }
+  return std::move(a);
+}
+
+// Combine all expressions from the container using &&.
+template <class container>
+Expr All(const container& c) {
+  Expr res;
+  for (const auto& e : c) {
+    if (res.get()) {
+      res = res && e;
+    } else {
+      res = e;
+    }
+  }
+  if (res.get()) {
+    return res;
+  } else {
+    return const_true();
+  }
+}
+
+// Create a select statement of the form cond ? on_true : 0
+Expr SelectElseZero(const Expr& cond, const Expr& on_true) {
+  return Select::make(cond, on_true, make_zero(on_true.type()));
+}
+
+// Simplify the expression as thoroughly as possible by using all available simplifiers.
+Expr SuperSimplify(Expr e, const Map<Var, Range>& vranges = Map<Var, Range>()) {
+  // For some reason no simplifier can detect that there is only one value of the variable
+  std::unordered_map<const Variable*, Expr> vmap;
+  for (const auto& var_range : vranges) {
+    if (is_const_int(var_range.second->extent, 1)) {
+      vmap[var_range.first.get()] = var_range.second->min;
+    }
+  }
+  if (!vmap.empty()) {
+    e = Substitute(e, vmap);
+  }
+
+  return CanonicalSimplify(Simplify(CanonicalSimplify(e, vranges), vranges), vranges);
```

Comment:
can you explain why call 3 times simplify here?

Reply:
It is simply the combination that gave the best result. I think I will rewrite this piece of code in the near future so that it doesn't use the Halide simplifier and uses the new simplifiers instead (and hopefully there won't be a need to make 3 calls).